### ISSUE TITLE
MSVC Errors when building Stokhos with CUDA enabled 

### ISSUE THREAD

**Issue by Ahdhn** [trilinos/Trilinos#1515]
I used MinGW to run a 'do-script' (shown below) that invokes cmake on windows. Cmake configured and generated the code successfully. However, building using VC always fails. I am interested in building _Stokhos_SPMVTest_ project only. When I tried to built it, the errors list shown below showed up. The system specs are listed below. Please advise. Thank you. 

**do-script**

```
#!/bin/sh
rm -f CMakeCache.txt;
rm -rf CMakeFiles
EXTRA_ARGS=$@
cmake \
-D CMAKE_INSTALL_PREFIX="C:/trilinos/install" \
-D CMAKE_BUILD_TYPE:STRING=RELEASE \
-D Trilinos_ENABLE_DEBUG=OFF \
-D Trilinos_ENABLE_CHECKED_STL=OFF \
-D Trilinos_ENABLE_EXPLICIT_INSTANTIATION=ON \
-D Trilinos_WARNINGS_AS_ERRORS_FLAGS="" \
-D Trilinos_ENABLE_ALL_PACKAGES:BOOL=OFF \
-D Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES=OFF \
-D Trilinos_DUMP_PACKAGE_DEPENDENCIES=OFF \
-D Trilinos_ENABLE_TESTS=ON \
-D Trilinos_ENABLE_OpenMP=ON \
-D Trilinos_ENABLE_CXX11=ON \
-D BUILD_SHARED_LIBS=ON \
-D DART_TESTING_TIMEOUT:STRING=600 \
-D CMAKE_CXX_COMPILER="C:/trilinos/nvcc_wrapper/nvcc_wrapper" \
-D CMAKE_C_COMPILER="C:/Program Files (x86)/Microsoft Visual Studio 12.0/VC/bin/cl.exe" \
-D TPL_BLAS_LIBRARIES="C:/open_blas/OpenBLAS-v0.2.19-Win64-int32/lib/libopenblas.dll.a" \
-D TPL_LAPACK_LIBRARIES="C:/open_blas/OpenBLAS-v0.2.19-Win64-int32/lib/libopenblas.dll.a" \
-D TPL_ENABLE_BinUtils=OFF \
-D TPL_ENABLE_Pthread=OFF \
-D TPL_ENABLE_CUDA=ON \
-D TPL_ENABLE_CUSPARSE=ON \
-D Trilinos_ENABLE_Stokhos=ON \
-D Stokhos_ENABLE_TESTS=ON \
-D Stokhos_ENABLE_EXAMPLES=ON \
-D Stokhos_ENABLE_EpetraExt=ON \
-D Stokhos_ENABLE_KokkosContainers=ON \
-D Kokkos_ENABLE_Cuda=ON \
-D Kokkos_ENABLE_Cuda_UVM=ON \
$EXTRA_ARGS \
../trilinos_source 2>&1 | tee -a configure.out
```

**Error List** 
```
                 DESCRIPTION                                                                                                  FILE                                                             LINE      COLUMN         PROJECT
error C2065: 'T' : undeclared identifier	                                                            C:\trilinos\trilinos_source\packages\kokkos\core\src\impl\Kokkos_Utilities.hpp     	61	   1	     Stokhos_SPMVTest
error C2143: syntax error : missing ';' before '&&'	                                                    C:\trilinos\trilinos_source\packages\kokkos\core\src\impl\Kokkos_Utilities.hpp     	67	   1	     Stokhos_SPMVTest
error C2144: syntax error : 'int' should be preceded by ';'	                                            C:\trilinos\trilinos_source\packages\kokkos\core\src\impl\Kokkos_Utilities.hpp     	67	   1	     Stokhos_SPMVTest
error C2146: syntax error : missing ';' before identifier '__host__'	                                C:\trilinos\trilinos_source\packages\kokkos\core\src\impl\Kokkos_Utilities.hpp     	65	   1	     Stokhos_SPMVTest
error C2146: syntax error : missing ';' before identifier 'T'	                                        C:\trilinos\trilinos_source\packages\kokkos\core\src\impl\Kokkos_Utilities.hpp     	67	   1	     Stokhos_SPMVTest
error C2923: 'std::remove_reference' : 'T' is not a valid template type argument for parameter '_Ty'	C:\trilinos\trilinos_source\packages\kokkos\core\src\impl\Kokkos_Utilities.hpp     	67	   1	     Stokhos_SPMVTest
error C3646: 'noexcept' : unknown override specifier	                                                C:\trilinos\trilinos_source\packages\kokkos\core\src\impl\Kokkos_Utilities.hpp     	68	   1	     Stokhos_SPMVTest
error C4430: missing type specifier - int assumed. Note: C++ does not support default-int	            C:\trilinos\trilinos_source\packages\kokkos\core\src\impl\Kokkos_Utilities.hpp     	68	   1	     Stokhos_SPMVTest
```
**System Specs:**

- Windows 7 - x64
- MSVC 2013
- NVIDIA GeForce GTX 970
- CUDA 7.5

In the nvcc wrapper, the only change made was to specify the host compiler as follows 
```
#host_compiler="C:/Program Files (x86)/Microsoft Visual Studio 12.0/VC/bin/cl.exe"
```

**Comment by Ahdhn** [comment]
I believe this due to the incomplete support of C++11 in VS 2013 as explained [here](https://msdn.microsoft.com/en-us/library/hh567368.aspx) specially with _constexpr_ and _noexcept_. Same thing was further confirmed in this discussion [here](https://github.com/kokkos/kokkos/issues/9).

Before closing this issue, I would like to ask if there is a way to build and run this project (_Stokhos_SPMVTest_) on Windows (e.g., using Cygwin) knowing that it will need Cuda. 

Thank you!!

**Comment by ibaned** [comment]
@mplooster @trilinos/stokhos

**Comment by mplooster** [comment]
@ibaned I'm afraid I can't comment very much at this point. I've never been able to build a package using MSVC that has a dependency on Kokkos, and I am not very familiar with stokhos. Sorry. In theory, it should be possible to build using Cygwin or MinGW, but having never tried myself I can't confirm (perhaps the stokhos package developers can comment more on that?). My best advice is to simply try it and see what happens.

**Comment by mhoemmen** [comment]
@Ahdhn It's been a while since I heard about Windows support, but I suspect you will need at least revision 3 of MSVC 2015.  That's when they actually finished implementing the C++98 standard (seriously; they didn't have a complete SFINAE implementation before).  This isn't even for CUDA; MSVC itself, without CUDA, was not a correct C++ implementation until that version.  (I am painfully familiar with older versions of MSVC and their C++ quirks.)

**Comment by mhoemmen** [comment]
@Ahdhn If you have a chance to try the most recent MSVC possible, without CUDA, please post your results on this issue.  Thanks!

**Comment by Ahdhn** [comment]
I tried with VS2017 but still I am getting few errors and it did not compile. I gave it a shot with cygwin (with cuda enabled). However, cygwin actually uses VS compiler (it asks for _cl.exe_ if it is not present in the PATH). I guess this is because it replies on the cuda toolkit that is installed on the host Windows.  

I am not running the test remotely on Linux machine. Thank you all for your help and fast responses.
